Release new version

nbirillo opened the pull request:
1. Add penalty algorithm (if you don't send history, it does not work, but the output JSON has a new field in all cases)
2. Merge subcategories into five main categories (to disable this feature you shout use `‑‑with‑all‑categories` flag)
3. Add difficulty level for each issue (**Note, if you use this flag, the output format will be changed.** To use this feature you should use `‑‑group‑by‑difficulty` flag. README file contains ax example of this format)
4. Some code refactoring
5. Update grade thresholds according to the empirical analysis (discussed with Kolya)
6. Update lintes for Java, Kotlin and Python
7. Support Java15

nbirillo commented:
@vyahhi @andrvb could you please check this PR?

andrvb commented:
Are file like `test/resources/evaluation/xlsx_target_files/target_sorted_order.xlsx` and similar files in `evaluation` dir needed?

nbirillo commented:
@andrvb I deleted unused inspectors (since they are too slow and we don't plan to use them in the tool) Also I deleted all inspectors sources since it can violate licenses. And of course, I deleted unnecessary resources.

In GitHub actions I have to use my own docker image since an intermediate version is not pished into stepik docker hub and does not contain all necessary changes (I built my docker image by Dockerfile in the repo)

Sorry for the big PR (I decreased it as it was possible). But in this PR we updated linters and it means we have to change configs and issues categorization. Several configs are too big, also we have some additional big files (like whitelists for flake8 in the repo and in the tool).

You can ignore the following files:
whitelist.txt (both of them) 
issue_types.py (for all inspectors) 
xml configs for linters 

It is about 1200 new lines of code. Also since we added  ‑‑group‑by‑difficulty and ‑‑with‑all‑categories we changed a lot of tests, it also takes a lot of new code lines.

vyahhi commented:
![source](https://user-images.githubusercontent.com/175030/133151696-72234ee3-b259-49bf-958b-4b37f54093b0.gif)

vyahhi commented:
@andrvb let's merge this PR?

andrvb commented on Dockerfile:
```diff
@@ -9,4 +9,34 @@ RUN pip install --no-cache-dir \
+# Install PMD
+RUN curl -sSLO https://github.com/pmd/pmd/releases/download/pmd_releases/${PMD_VERSION}/pmd-bin-${PMD_VERSION}.zip \
+    && unzip pmd-bin-${PMD_VERSION}.zip -d ${PMD_DIRECTORY}
+
```
It would be better to move the linters installation before the hyperstyle code installation. This will allow you to use the Docker cache because the hyperstyle code changes more frequently than the linter versions.
See https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache

Or (maybe it even more better) move the linters installation to base docker image. In this case, you can continue to use the base image in GitHub actions.

nbirillo replied:
Should I do it in this PR? Or in another one?``

andrvb replied:
@nbirillo, whatever works better for you.

nbirillo replied:
Let's do it in the next PR

vyahhi commented:
OK, I'm merging this PR since it has no major issues to fix anymore.
